forch/faucet_state_collector.py

```diff
@@ -823,6 +823,23 @@ def process_port_learn(self, timestamp, name, port, mac, ip_addr):
                 devices_placement = DevicePlacement(switch=name, port=port, connected=True)
                 self._placement_callback(mac, devices_placement)
 
+    @_dump_states
+    def process_port_expire(self, timestamp, name, port, mac):
+        """process port expire event"""
+        with self.lock:
+            LOGGER.info('Learned entry %s at %s:%s expired.', mac, name, port)
+            learned_macs = self.switch_states[name][LEARNED_MACS]
+            if mac in learned_macs:
+                learned_macs.remove(mac)
+            else:
+                LOGGER.warning('Entry %s doesnt exist in learned macs set', mac)
+            is_last_mac_entry = True
```

Comment:
Much cleaner to be direct about it, rather than indirectly calculating when to remove it -- learned_macs contains a map of switches, so each time through you remove the switch that expired for this mac, and then if that map is empty you can delete the entire entry.

Reply:
yup. Done.